src/ai/defaultai.cc

```diff
@@ -6754,9 +6873,209 @@ void DefaultAI::update_player_stat(const Time& gametime) {
+					inputs[24] = cur_land >= player_statistics.get_max_land() ? -4 : 1;
+					inputs[25] = player_statistics.get_diplo_score(j) > 0 ? 5 : -1;
+					inputs[26] = player_statistics.get_diplo_score(j) > 5 ? 3 : -2;
+					inputs[27] = cur_strength >= player_statistics.get_max_power() ?
+                               -(std::abs(management_data.get_military_number_at(197)) / 10) :
+                               (std::abs(management_data.get_military_number_at(197)) / 10);
+					inputs[28] = cur_land >= player_statistics.get_max_land() ? -7 : 2;
+					inputs[29] = gametime < Time((30 + RNG::static_rand(20)) * 60 * 1000) ? -5 : 0;
+					inputs[30] = gametime < Time((60 + RNG::static_rand(30)) * 60 * 1000) ? -3 : 0;
+					inputs[31] = cur_strength < player_statistics.get_max_power() &&
+					                   player_statistics.get_max_power() < cur_strength + me_strength &&
+					                   me_strength < player_statistics.get_max_power() ?
+                               8 :
+                               0;
+					inputs[32] =
+					   cur_strength * 3 + cur_land < player_statistics.get_max_power() * 3 +
+					                                    player_statistics.get_max_land() &&
+					         player_statistics.get_max_power() * 3 + player_statistics.get_max_land() <
+					            (cur_strength + me_strength) * 3 + cur_land + me_land ?
+                     5 :
+                     -5;
+					inputs[33] =
+					   cur_strength * 2 + cur_land < player_statistics.get_max_power() * 2 +
+					                                    player_statistics.get_max_land() &&
+					         player_statistics.get_max_power() * 2 + player_statistics.get_max_land() <
+					            (cur_strength + me_strength) * 2 + cur_land + me_land ?
+                     4 :
+                     0;
+					inputs[34] = this_player->team_number() == me->team_number() ? 5 : -5;
+					inputs[35] = this_player->team_number() == me->team_number() ? 3 : -3;
+					inputs[36] = player_statistics.get_diplo_score(j) > 10 ? 2 : -2;
+					inputs[37] = player_statistics.get_is_enemy(j) &&
+					                   player_statistics.player_seen_lately(j, gametime) &&
+					                   !player_statistics.strong_enough(pn) ?
+                               8 :
+                               -2;
+					inputs[38] = player_statistics.get_is_enemy(j) &&
+					                   player_statistics.player_seen_lately(j, gametime) &&
+					                   !player_statistics.strong_enough(pn) ?
+                               5 :
+                               0;
+					inputs[39] = player_statistics.strong_enough(pn) ? -10 : 2;
+					inputs[40] = player_statistics.strong_enough(pn) ? -5 : 2;
+					inputs[41] = player_statistics.get_is_enemy(j) &&
+					                   player_statistics.player_seen_lately(j, gametime) &&
+					                   player_statistics.enemies_seen_lately_count(gametime) > 1 ?
+                               5 :
+                               0;
```

Comment:
These checks with is_enemy are major contributors to the swings on join/leave.

Reply:
Ok I will reduce the gap between the 2 values as suggested. However please keep in mind, that not all the inputs are evaluated. Only the inputs that are chosen to be valuable based on AI training values.      
So this whole thing is subject to AI Training as well.

Reply:
Actually my suggestion now is edit2 above. Keep the bigger gap, because it's an important factor, but the part whether we're allied is less important for this factor.

Unfortunately I don't know how the AI training works. How do you set a target function for a thing as complex as diplomacy?

Reply:
Well, usually training has been done by @tibor95. I think we can't evaluate diplomacy directly but might rely on the outcome (i.e. winning the game).

Reply:
made the enemy status a factor